### DISCUSSION TITLE
`listen` AsyncNotifierProvider and awaiting for data?

### DISCUSSION THREAD

**Discussion by tonnyavery** [rrousselGit/riverpod#2267] [Q&A] [answered]
I find AsyncValue introduces a lot of confusion to the whole API.

With `AsyncNotifierProvider`, running `ref.listen` on it results in both `previous` and `next` be of AsyncValue.
When testing, I often get that `next` is already resolved to value, but `previous` is still in AsyncLoading state.
Could it be related to https://github.com/rrousselGit/riverpod/discussions/1772?

How can I make sure both `previous` and `next` are resolved into data to compare them and decide on my listener logic?

**Comment by rrousselGit** [comment]
That doesn't make sense, because "loading" is a valid state too. People also want to know when an item changes from/to loading

But AsyncValue exposes the previous state too. You can read that.
You should be able to do `previous?.value` / `previous?.error`

**Reply by tonnyavery** [reply]
My expectation is that running `ref.listen(asyncNotifierProvider)` I will get previous and next value, compare them and run my logic.
What seems to happen is, I get previous state in AsyncLoading and next state in AsyncValue.

If initial state of my asyncNotifierProvider is empty, adding new entries, I expect
- after first item added, listen callback should get `previous=empty`, `next=my new entry 1`
- after second item added, listen callback should get `previous=my new entry 1`, `next=my new entry 2`
- and so on

What I get instead is
- first item added, `previous=AsyncLoaded`, next=`My new entry1`
- second item added, `previous=AsyncLoaded`, next=`My new entry2`

**Reply by tonnyavery** [reply]
I think I see what the problem is.
If `previous, next` is not about values, but about states, then if adding to the listened provider is very fast, some states get missing.
So instead of the path of `previous=Loading, next=Empty`, `previous=Empty, next=Loading`, `previous=Loading, next=Data1` and so on, I just get `previous=Data1, next=Data1`, just because AsyncValue happened to resolve into data already.
That kind of removes the possibility to compare the data added to AsyncNotifier and decide on the logic when data really changes.
And those missing states in between do not make it simple to workaround.

**Reply by tonnyavery** [reply]
This still sounds wrong though.
I am not delivering AsyncLoading from my AsyncNotifier. So why is there ever a case when previous == next?

**Reply by rrousselGit** [reply]
I don't understand your problem. What prevents you from doing this:

```dart
if (previous?.value != next.value) {
 }
```

**Reply by tonnyavery** [reply]
I will try to come up with a simple code example.
From my tests, this evaluation is never `true` and is always `false` because somehow previous always equals next.

**Reply by rrousselGit** [reply]
That's not the case.

Going from AsyncData(42) => AsyncLoading() => AsyncData(21) will cause prev.value/next.value to be different when going from loading to AsyncData(21)

**Reply by rrousselGit** [reply]
You could also check `asyncValue.isLoading` is that's your thing.

**Reply by tonnyavery** [reply]
@rrousselGit Here is a small sample that reproduces the problem: https://dartpad.dev/?id=a9727def4fb01c4514a3e7d8473683e8

Run it in dartpad and open console.
Press on + and see that both previous and next always match each other.

**Reply by tonnyavery** [reply]
I understand that the argument could be that I didn't send `AsyncValue.loading()` before I have started `guard()`, but that still doesn't explain why previous and next values do match on every step.

And generally speaking, AsyncValue implies a bit too much how I should handle the state. What if my Loading goes on top of the Data? Or I want to have both Data and Error at the same time?
Is there any way to use Future/Async providers without AsyncValue? I would like to have more control over what is `state` exactly.

**Reply by rrousselGit** [reply]
The problem is unrelated. You mutated the list. Of course the list reference will be the same.
If you want == to be different, clone the list instead of mutating it.

**Reply by tonnyavery** [reply]
Wouldn't it ruin the whole point of a `local cache` or `local database`? Assuming database can grow over time, cloning the data for every mutation will be very expensive. Which is why I use `add`, and not `[...originalList, newItem]`

**Reply by tonnyavery** [reply]
The gist example is simplified, in my code I don't have a list, I have a Map of Lists, which models local version of a future coming database. Cloning the whole "database snapshot" for every mutation doesn't make any sense.
Am I using some kind of wrong provider to implement this using riverpod?

**Reply by rrousselGit** [reply]
This has nothing to do with Riverpod. You're expecting the list equality to change, but you're mutating it instead of making of a new instance.
It's a paradox.

Either don't rely on == or don't mutate the list.

**Reply by tonnyavery** [reply]
Changing to ` Future<List<int>> _readDatabase() async => [...values];` in the gist resolve the issue, thanks for clarification.

My question about ways to use Riverpod Future/Async without AsyncValue still stands as I find it brings a lot of confusion the process. Like this issue with the state.
In the code it looks like I assign the new state by setting
```state = await AsyncValue.guard(() {
  return myList;
})```
While in practice this ends up using the same list reference even though state was "reassigned".

Comparing this to the reducers in Redux, mistake would have been obvious from the code:

`reducer(action, state) => { list: myList }` instead of `reducer(action, state) => { ...state, list: myList }`

**Reply by rrousselGit** [reply]
Redux would have exactly the same issue if you mutated the list.

Redux asks you to clone the list instead of mutating it

**Reply by tonnyavery** [reply]
I am not saying that Redux works differently. My point is that due to the existence of AsyncValue, it is easier to make the mistake I made in Riverpod because it does look like I have correctly changed the state when assigning `state = AsyncValue.guard`, while in reality I haven't really.

**Comment by tonnyavery** [comment, marked as answer]
The problem was in reusing list instance when setting state in AsyncNotifier.
As it was not the copy, list was being updated instantly and previous/next wasn't working correctly. See the thread for details and sample
